simulator/infra/ports/t/topology_test.cpp

```diff
@@ -0,0 +1,147 @@
+/**
+ * Topology unit test
+ * @author Eric Konks
+ */
+
+#include "../module.h"
+#include <catch.hpp>
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/json_parser.hpp>
+
+namespace pt = boost::property_tree;
+
+static auto read_json_from_file( const std::string& path)
+{
+    pt::ptree t;
+    read_json( TEST_PATH + path, t);
+    return t;
+}
+
+struct SomeTopology : public Root
+{
+    struct A : public Module {
+        std::unique_ptr<WritePort<int>> to_C;
+        std::unique_ptr<WritePort<int>> to_D;
+        std::unique_ptr<ReadPort<int>> from_C;
+        std::unique_ptr<ReadPort<int>> from_D;
+        explicit A( Module* root) : Module( root, "A")
+        {
+            to_C = make_write_port<int>( "A_to_C", PORT_BW);
+            to_D = make_write_port<int>( "A_to_D", PORT_BW);
+            from_C = make_read_port<int>( "C_to_A", PORT_LATENCY);
+            from_D = make_read_port<int>( "D_to_A", PORT_LATENCY);
+        }
+    };
+    struct B : public Module {
+        std::unique_ptr<WritePort<int>> to_C;
+        std::unique_ptr<WritePort<int>> to_D;
+        std::unique_ptr<ReadPort<int>> from_C;
+        std::unique_ptr<ReadPort<int>> from_D;
+        explicit B( Module* root) : Module( root, "B")
+        {
+            to_C = make_write_port<int>( "B_to_C", PORT_BW);
+            to_D = make_write_port<int>( "B_to_D", PORT_BW);
+            from_C = make_read_port<int>( "C_to_B", PORT_LATENCY);
+            from_D = make_read_port<int>( "D_to_B", PORT_LATENCY);
+        }
+    };
+    struct C : public Module {
+        std::unique_ptr<WritePort<int>> to_A;
+        std::unique_ptr<WritePort<int>> to_B;
+        std::unique_ptr<ReadPort<int>> from_A;
+        std::unique_ptr<ReadPort<int>> from_B;
+        explicit C( Module* root) : Module( root, "C")
+        {
+            to_A = make_write_port<int>( "C_to_A", PORT_BW);
+            to_B = make_write_port<int>( "C_to_B", PORT_BW);
+            from_A = make_read_port<int>( "A_to_C", PORT_LATENCY);
+            from_B = make_read_port<int>( "B_to_C", PORT_LATENCY);
+        }
+    };
+    struct D : public Module {
+        std::unique_ptr<WritePort<int>> to_A;
+        std::unique_ptr<WritePort<int>> to_B;
+        std::unique_ptr<ReadPort<int>> from_A;
+        std::unique_ptr<ReadPort<int>> from_B;
+        explicit D( Module* root) : Module( root, "D")
+        {
+            to_A = make_write_port<int>( "D_to_A", PORT_BW);
+            to_B = make_write_port<int>( "D_to_B", PORT_BW);
+            from_A = make_read_port<int>( "A_to_D", PORT_LATENCY);
+            from_B = make_read_port<int>( "B_to_D", PORT_LATENCY);
+        }
+    };
+    std::unique_ptr<A> a;
+    std::unique_ptr<B> b;
+    std::unique_ptr<C> c;
+    std::unique_ptr<D> d;

```

Comment:
You can just have `A a` here without a pointer, that would be a little bit cleaner.